### ISSUE TITLE
AL-Validation Error: "An exception occurred during a WebClient request."

### ISSUE THREAD

**Issue by bernalfr** [microsoft/navcontainerhelper#2486]
PLEASE DO NOT INCLUDE ANY PASSWORDS OR TOKENS IN YOUR ISSUE!!!

**Describe the issue**
The error: "An exception occurred during a WebClient request." is thrown by the AL-Validation tool as it tries to determine/download artifact information.  Multiple execution attempts are required to have the tool finish gathering information and proceed with the rest of the validation process.

**Scripts used to create container and cause the issue**
```
        $validationResult = Run-AlValidation `
            -containerName $containerName `
            -installApps @($dependencies) `
            -previousApps @($previousApps) `
            -apps @($apps) `
            -validateCurrent `
            -sastoken $sasToken `
            -countries $countries `
            -affixes "LAX" `
            -supportedCountries $countries `
            -skipVerification `
            -verbose
```

**Full output of scripts**
```
VERBOSE: Perform operation 'Enumerate CimInstances' with following parameters, ''namespaceName' = root\cimv2,'className' = Win32_Operating
System'.
VERBOSE: Operation 'Enumerate CimInstances' complete.
VERBOSE: POST https://marketplace.visualstudio.com/_apis/public/gallery/extensionquery?api-version=3.0-preview.1 with -1-byte payload
VERBOSE: received 4428-byte response of content type application/json; charset=utf-8; api-version=3.0-preview.1
  _____                               _                
 |  __ \                             | |               
 | |__) |_ _ _ __ __ _ _ __ ___   ___| |_ ___ _ __ ___ 
 |  ___/ _` | '__/ _` | '_ ` _ \ / _ \ __/ _ \ '__/ __|
 | |  | (_| | | | (_| | | | | | |  __/ |_  __/ |  \__ \
 |_|   \__,_|_|  \__,_|_| |_| |_|\___|\__\___|_|  |___/

Container name               AL-VALIDATE-APP
Credential                   admin/Hygy9287
MemoryLimit                  8G
validateVersion              
validateCurrent              True
validateNextMinor            False
validateNextMajor            False
SasToken                     Specified
countries                    us,au,ca,gb,nz
validateCountries            us,au,ca,gb,nz
affixes                      FOO
supportedCountries           us,au,ca,gb,nz
vsixFile                     
Install Apps
- DependencyApp01
- DependencyApp02
- DependencyApp03
Previous Apps
- PreviousApp01
Apps
- CurrentApp01

  _____       _ _ _                                          _        _                            
 |  __ \     | | (_)                                        (_)      (_)                           
 | |__) |   _| | |_ _ __   __ _    __ _  ___ _ __   ___ _ __ _  ___   _ _ __ ___   __ _  __ _  ___ 
 |  ___/ | | | | | | '_ \ / _` |  / _` |/ _ \ '_ \ / _ \ '__| |/ __| | | '_ ` _ \ / _` |/ _` |/ _ \
 | |   | |_| | | | | | | | (_| | | (_| |  __/ | | |  __/ |  | | (__  | | | | | | | (_| | (_| |  __/
 |_|    \__,_|_|_|_|_| |_|\__, |  \__, |\___|_| |_|\___|_|  |_|\___| |_|_| |_| |_|\__,_|\__, |\___|
                           __/ |   __/ |                                                 __/ |     
                          |___/   |___/                                                 |___/      

Pulling mcr.microsoft.com/businesscentral:10.0.17763.2803-filesonly-dev

Pulling generic image took 1 seconds
  _____       _                      _                         _   _  __           _       
 |  __ \     | |                    (_)                       | | (_)/ _|         | |      
 | |  | | ___| |_ ___ _ __ _ __ ___  _ _ __   ___    __ _ _ __| |_ _| |_ __ _  ___| |_ ___ 
 | |  | |/ _ \ __/ _ \ '__| '_ ` _ \| | '_ \ / _ \  / _` | '__| __| |  _/ _` |/ __| __/ __|
 | |__| |  __/ |_  __/ |  | | | | | | | | | |  __/ | (_| | |  | |_| | || (_| | (__| |_\__ \
 |_____/ \___|\__\___|_|  |_| |_| |_|_|_| |_|\___|  \__,_|_|   \__|_|_| \__,_|\___|\__|___/
                                                                                           
VERBOSE: DownloadString https://bcartifacts.blob.core.windows.net/sandbox/?comp=list&restype=container
VERBOSE: DownloadString https://bcartifacts.blob.core.windows.net/sandbox/?comp=list&restype=container&marker=2!84!MDAwMDE5ITE4LjIuMjYyMTc
uMzI0MTgvcGwhMDAwMDI4ITE2MDEtMDEtMDFUMDA6MDA6MDAuMDAwMDAwMFoh
VERBOSE: DownloadString https://bcartifacts.blob.core.windows.net/sandbox/?comp=list&restype=container&marker=2!84!MDAwMDE5ITE5LjEuMzE4ODY
uMzQ2NDYvcm8hMDAwMDI4ITk5OTktMTItMzFUMjM6NTk6NTkuOTk5OTk5OVoh
VERBOSE: DownloadString https://bcartifacts.blob.core.windows.net/sandbox/?comp=list&restype=container&marker=2!84!MDAwMDE5ITE5LjEuMzE4ODY
uMzQ2NzQvZWUhMDAwMDI4ITE2MDEtMDEtMDFUMDA6MDA6MDAuMDAwMDAwMFoh
Run-AlValidation Telemetry Correlation Id: aee3fc90-e67d-4083-a2db-945525c20466
Exception calling "DownloadString" with "1" argument(s): "An exception occurred during a WebClient request."
At C:\Program Files\WindowsPowerShell\Modules\BcContainerHelper\3.0.9\Misc\Get-BcArtifactUrl.ps1:197 char:13
+             $Response = $webClient.DownloadString("$GetListUrl$nextMa ...
+             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [], ParentContainsErrorRecordException
    + FullyQualifiedErrorId : WebException
```

**Screenshots**

![Screen Shot 2022-05-19 at 23 25 16](https://user-images.githubusercontent.com/35543556/169458359-234fc6e0-f927-4422-b663-90c56155ae9f.png)

**Additional context**
- The issue occurs with the AL-Validate tool in bccontainerhelper versions 3.0.8 and 3.0.9-preview606
- The issue has been occurring throughout this past week.

**Comment by bszentesi** [comment]
I have the same error message when run the Sign-BCContainerApp cmdlet. The error occur when I try to download the tool with this command: (New-Object System.Net.WebClient).DownloadFile($winSdkSetupUrl, $winSdkSetupExe)
I am using prestaged images, so hopefully artifact download won't cause an issue for me :)

I have implemented a workaround by add a retry logic to the Run-AlPipeline's parameter list:
            -SignBcContainerApp { 
                Param([Hashtable]$parameters)
                Invoke-ScriptInBcContainer -containerName $containerName -ScriptBlock {
                    if (!(Test-Path "C:\Program Files (x86)\Windows Kits\10\bin\*\x64\SignTool.exe")) {
                        Write-Host "Prestage - Downloading Signing Tools"
                        $winSdkSetupExe = "c:\run\install\winsdksetup.exe"
                        $winSdkSetupUrl = "https://go.microsoft.com/fwlink/p/?LinkID=2023014"
                        New-Item -ItemType Directory -Force -Path $winSdkSetupExe.replace('\winsdksetup.exe','') -ErrorAction SilentlyContinue
                        $retries = 5
                        $secondsDelay = 1
                        $temp = $ErrorActionPreference
                        $ErrorActionPreference = "Stop"
                        $retrycount = 0
                        $completed = $false
                        while (-not $completed) {
                            try {
                                (New-Object System.Net.WebClient).DownloadFile($winSdkSetupUrl, $winSdkSetupExe)
                                Write-Host ("Command [{0}] succeeded." -f $command)
                                $completed = $true
                            } catch {
                                if ($retrycount -ge $retries) {
                                    Write-Host ("Command [{0}] failed the maximum number of {1} times." -f $command, $retrycount)
                                    throw
                                } else {
                                    Write-Host ("Command [{0}] failed. Retrying in {1} seconds." -f $command, $secondsDelay)
                                    Start-Sleep $secondsDelay
                                    $retrycount++
                                }
                            }
                        }
                        $ErrorActionPreference = $temp
                        Write-Host "Prestage - Installing Signing Tools"
                        Start-Process $winSdkSetupExe -ArgumentList "/features OptionId.SigningTools /q" -Wait
                        if (!(Test-Path "C:\Program Files (x86)\Windows Kits\10\bin\*\x64\SignTool.exe")) {
                            throw "Prestage - Cannot locate signtool.exe after installation"
                        }
                    }
                }
                
                Sign-BcContainerApp @parameters
            }

**Comment by freddydk** [comment]
Will see if I can add some resilience to this

**Comment by bernalfr** [comment]
Hello.

It has been a while since I last had to use the AL validation tool.  I had to use it today to verify an updated extension and ran into a similar error.  Uninstalling the existing version of BcContainerHelper (3.0.8) and installing the most recent build (3.0.11-preview624) led to the same result.  Repeated attempts to run the tool also resulted in an error.

Sample Run-AL-Validation output:

    VERBOSE: Perform operation 'Enumerate CimInstances' with following parameters, ''namespaceName' = root\cimv2,'className' = Win32_OperatingSystem'.
    VERBOSE: Operation 'Enumerate CimInstances' complete.
    VERBOSE: POST https://marketplace.visualstudio.com/_apis/public/gallery/extensionquery?api-version=3.0-preview.1 with -1-byte payload
    VERBOSE: received 4408-byte response of content type application/json; charset=utf-8; api-version=3.0-preview.1
      _____                               _                
     |  __ \                             | |               
     | |__) |_ _ _ __ __ _ _ __ ___   ___| |_ ___ _ __ ___ 
     |  ___/ _    | '__/ _    | '_     _ \ / _ \ __/ _ \ '__/ __|
     | |  | (_| | | | (_| | | | | | |  __/ |_  __/ |  \__ \
     |_|   \__,_|_|  \__,_|_| |_| |_|\___|\__\___|_|  |___/
    
    Container name                  AL-VALIDATE-EXT
    Credential                      admin/Niku0570
    MemoryLimit                     8G
    validateVersion                 
    validateCurrent                 True
    validateNextMinor               False
    validateNextMajor               False
    SasToken                        Specified
    countries                       us
    validateCountries               us
    affixes                         AFX
    supportedCountries              us
    ObsoleteTagMinAllowedMajorMinor 
    vsixFile                        
    Install Apps
    - app1.app
    - app2.app
    - app3.app
    Previous Apps
    - prevapp1.app
    Apps
    - currapp1.app
    
      _____       _ _ _                                          _        _                            
     |  __ \     | | (_)                                        (_)      (_)                           
     | |__) |   _| | |_ _ __   __ _    __ _  ___ _ __   ___ _ __ _  ___   _ _ __ ___   __ _  __ _  ___ 
     |  ___/ | | | | | | '_ \ / _    |  / _    |/ _ \ '_ \ / _ \ '__| |/ __| | | '_     _ \ / _    |/ _    |/ _ \
     | |   | |_| | | | | | | | (_| | | (_| |  __/ | | |  __/ |  | | (__  | | | | | | | (_| | (_| |  __/
     |_|    \__,_|_|_|_|_| |_|\__, |  \__, |\___|_| |_|\___|_|  |_|\___| |_|_| |_| |_|\__,_|\__, |\___|
                               __/ |   __/ |                                                 __/ |     
                              |___/   |___/                                                 |___/      
    
    Pulling mcr.microsoft.com/businesscentral:10.0.17763.2928-filesonly-dev
    
    Pulling generic image took 0 seconds
      _____       _                      _                         _   _  __           _       
     |  __ \     | |                    (_)                       | | (_)/ _|         | |      
     | |  | | ___| |_ ___ _ __ _ __ ___  _ _ __   ___    __ _ _ __| |_ _| |_ __ _  ___| |_ ___ 
     | |  | |/ _ \ __/ _ \ '__| '_     _ \| | '_ \ / _ \  / _    | '__| __| |  _/ _    |/ __| __/ __|
     | |__| |  __/ |_  __/ |  | | | | | | | | | |  __/ | (_| | |  | |_| | || (_| | (__| |_\__ \
     |_____/ \___|\__\___|_|  |_| |_| |_|_|_| |_|\___|  \__,_|_|   \__|_|_| \__,_|\___|\__|___/
                                                                                               
    VERBOSE: DownloadString https://bcartifacts.blob.core.windows.net/sandbox/?comp=list&restype=container
    VERBOSE: DownloadString https://bcartifacts.blob.core.windows.net/sandbox/?comp=list&restype=container&marker=2!84!MDAwMDE5ITE4LjIuMjYyMTcuMzI0MTgvcGwh
    MDAwMDI4ITE2MDEtMDEtMDFUMDA6MDA6MDAuMDAwMDAwMFoh
    VERBOSE: DownloadString https://bcartifacts.blob.core.windows.net/sandbox/?comp=list&restype=container&marker=2!84!MDAwMDE5ITE5LjEuMzE4ODYuMzM2MzgvZWUh
    MDAwMDI4ITk5OTktMTItMzFUMjM6NTk6NTkuOTk5OTk5OVoh
    VERBOSE: DownloadString https://bcartifacts.blob.core.windows.net/sandbox/?comp=list&restype=container&marker=2!84!MDAwMDE5ITE5LjEuMzE4ODYuMzQ2NzQvZWUh
    MDAwMDI4ITE2MDEtMDEtMDFUMDA6MDA6MDAuMDAwMDAwMFoh
    Run-AlValidation Telemetry Correlation Id: e1633657-090f-461a-b33d-d71d5c5feb65
    Exception calling "DownloadString" with "1" argument(s): "An exception occurred during a WebClient request."
    At C:\Program Files\WindowsPowerShell\Modules\BcContainerHelper\3.0.11\Misc\Get-BcArtifactUrl.ps1:197 char:13
    +             $Response = $webClient.DownloadString("$GetListUrl$nextMa ...
    +             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
        + CategoryInfo          : NotSpecified: (:) [], ParentContainsErrorRecordException
        + FullyQualifiedErrorId : WebException

**Comment by bernalfr** [comment]
I was able to apply a workaround by modifying Get-BcArtifactUrl:

            $webclient = New-Object System.Net.WebClient
            $Artifacts = @()
            $nextMarker = ""
    
            #ZZ-MOD >>
            $currentMarker = ''
            $downloadAttempt = 0
            $downloadRetryAttempts = 10
            #ZZ-MOD <<
            
            do {
                #ZZ-MOD >>
                if ($currentMarker -ne $nextMarker)
                {
                    $currentMarker = $nextMarker
                    $downloadAttempt = 1
                }
                #ZZ-MOD <<
                
                Write-Verbose "DownloadString $GetListUrl$nextMarker"

                #ZZ-MOD >>
                #$Response = $webClient.DownloadString("$GetListUrl$nextMarker")
                #$enumerationResults = ([xml]$Response).EnumerationResults

                try
                {
                    $Response = $webClient.DownloadString("$GetListUrl$nextMarker")
                    $enumerationResults = ([xml]$Response).EnumerationResults
                }
                catch
                {
                    $downloadAttempt += 1

                    Write-Host "WARNING: The following error occurred:"
                    Write-Host -ForegroundColor Red $_
                    Write-Host
                    
                    if ($downloadAttempt -le $downloadRetryAttempts)
                    {
                        Write-Host "Repeating download attempt (" $downloadAttempt.ToString() " of " $downloadRetryAttempts.ToString() ")..."
                        Write-Host
                    }
                    else
                    {
                        Write-Host Error Stack:
                        Write-Host -ForegroundColor Red $_.ScriptStackTrace
                        Write-Host

                        throw "Script has been terminated."
                    }                
                }
                #ZZ-MOD <<
                
                if ($enumerationResults.Blobs) {
                    if (($After) -or ($Before)) {
                        $artifacts += $enumerationResults.Blobs.Blob | % {
                            if ($after) {
                                $blobModifiedDate = [DateTime]::Parse($_.Properties."Last-Modified")
                                if ($before) {
                                    if ($blobModifiedDate -lt $before -and $blobModifiedDate -gt $after) {
                                        $_.Name
                                    }
                                }
                                elseif ($blobModifiedDate -gt $after) {
                                    $_.Name
                                }
                            }
                            else {
                                $blobModifiedDate = [DateTime]::Parse($_.Properties."Last-Modified")
                                if ($blobModifiedDate -lt $before) {
                                    $_.Name
                                }
                            }
                        }
                    }
                    else {
                        $artifacts += $enumerationResults.Blobs.Blob.Name
                    }
                }
                $nextMarker = $enumerationResults.NextMarker
                if ($nextMarker) {
                    $nextMarker = "&marker=$nextMarker"
                }
            } while ($nextMarker)

**Comment by freddydk** [comment]
Thanks, will add this to the next prerelease

**Comment by freddydk** [comment]
Shipped in BcContainerHelper 3.0.11
